### PR TITLE
Add support for .config file generation for test projects

### PR DESCRIPTION
**Pull request by pakrym**
https://github.com/dotnet/cli/issues/807
@victorhurdugaci  @davidfowl  @anurse

### CODE REVIEW

**Comment by victorhurdugaci** [comment]
Tests?

**Comment by pakrym** [comment]
Nope

**Comment by victorhurdugaci** [comment]
Why?

**Comment by victorhurdugaci** [comment]
Also, does it work on publish?

**Comment by pakrym** [comment]
Yes

File: src/Microsoft.DotNet.Tools.Compiler/Program.cs [hunk 1]
```diff
@@ -354,16 +356,22 @@ private static bool CompileProject(ProjectContext context, CompilerCommandApp ar
             }
             else if (!string.IsNullOrEmpty(context.ProjectFile.TestRunner))
```
Comment (davidfowl) [inline src/Microsoft.DotNet.Tools.Compiler/Program.cs]: Wow this is super hacky. Why does compile know about test runners.... We need to decouple that.

/cc @piotrpMSFT

**Comment by pakrym** [comment]
:up: :date: tests

File: test/TestProjects/TestLibraryWithRunner/NuGet.Config [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <packageSources>
+    <!--To inherit the global NuGet package sources remove the <clear/> line below -->
+    <clear />
+    <add key="dotnet-core" value="https://www.myget.org/F/dotnet-core/api/v3/index.json" />
```
Comment (victorhurdugaci) [inline test/TestProjects/TestLibraryWithRunner/NuGet.Config]: Is this file needed? If yes, do you really need all these feeds?

File: test/Microsoft.DotNet.Tools.Publish.Tests/Microsoft.DotNet.Tools.Publish.Tests.cs [hunk 3]
```diff
@@ -130,6 +130,30 @@ public void LibraryPublishTest()
...
+            RunRestore(testLibDir.Path);
+
+            var testProject = GetProjectPath(testLibDir);
+            var publishCommand = new PublishCommand(testProject);
+            publishCommand.Execute().Should().Pass();
+
+            publishCommand.GetOutputDirectory().Should().HaveFile("TestLibraryWithRunner.dll");
+            publishCommand.GetOutputDirectory().Should().HaveFile("TestLibraryWithRunner.pdb");
+            publishCommand.GetOutputDirectory().Should().HaveFile("TestLibraryWithRunner.deps");
+            publishCommand.GetOutputDirectory().Should().HaveFile("TestLibraryWithRunner.dll.config");
```
Comment (victorhurdugaci) [inline test/Microsoft.DotNet.Tools.Publish.Tests/Microsoft.DotNet.Tools.Publish.Tests.cs]: You should check the contents of this file because this test will pass even if the file is empty

File: test/TestProjects/TestLibraryWithRunner/NuGet.Config (hunk 2 above)
Comment (pakrym) [inline test/TestProjects/TestLibraryWithRunner/NuGet.Config]: I did the same way other tests are done

File: test/Microsoft.DotNet.Tools.Publish.Tests/Microsoft.DotNet.Tools.Publish.Tests.cs (hunk 3 above)
Comment (pakrym) [inline test/Microsoft.DotNet.Tools.Publish.Tests/Microsoft.DotNet.Tools.Publish.Tests.cs]: It's not file content logic test, this test check file layout

File: test/TestProjects/TestLibraryWithRunner/NuGet.Config (hunk 2 above)
Comment (victorhurdugaci) [inline test/TestProjects/TestLibraryWithRunner/NuGet.Config]: Remove all but nuget.org. It will make the test faster

File: test/Microsoft.DotNet.Tools.Publish.Tests/Microsoft.DotNet.Tools.Publish.Tests.cs (hunk 3 above)
Comment (victorhurdugaci) [inline test/Microsoft.DotNet.Tools.Publish.Tests/Microsoft.DotNet.Tools.Publish.Tests.cs]: Then add another test for the file content. I don't think there's one

Comment (pakrym) [inline test/Microsoft.DotNet.Tools.Publish.Tests/Microsoft.DotNet.Tools.Publish.Tests.cs]: Its unrelated to this isuue

File: src/Microsoft.DotNet.Tools.Compiler/Program.cs (hunk 1 above)
Comment (piotrpMSFT) [inline src/Microsoft.DotNet.Tools.Compiler/Program.cs]: discussed offline. Let's make this a first-class compiler option that's set true by default for test projects and executables.

**Comment by victorhurdugaci** [comment]
There's a failure on Ubuntu and OSX. Could be related:

```
09:30:59    Microsoft.DotNet.Tools.Publish.Tests.PublishTests.TestLibraryPublishTest [FAIL]
09:30:59       Expected command to pass but it exited with 1.
09:30:59 Executing - dotnet restore --quiet /tmp/DotnetCLITests/2f24c868-5288-4c19-91f4-27d95c771080/TestLibrary
09:30:59       Stack Trace:
09:30:59             at FluentAssertions.Execution.LateBoundTestFramework.Throw(String message)
09:30:59             at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
09:30:59             at Microsoft.DotNet.Tools.Test.Utilities.CommandResultAssertions.Pass()
09:30:59             at Microsoft.DotNet.Tools.Publish.Tests.PublishTests.TestLibraryPublishTest()
09:31:00 Microsoft .NET Development Utility CoreClr-x64-1.0.0-rc1-16231
09:31:00 
```

Please investigate and then let's get this merged.

**Comment by pakrym** [comment]
Build agents don't have full .net reference assemblies, so It's unable to run publish for dnx451 target

**Comment by victorhurdugaci** [comment]
Then you probably have to disable the test on those platforms or talk to whoever maintains the agents to install mono. Otherwise every checkin after you merge this will be broken.

**Comment by pakrym** [comment]
Yep, working on it
